core/services/blockhashstore/feeder.go

```diff
@@ -0,0 +1,164 @@
+	// For VRF V2, it will be an integer in string form.
+	ID string
+
+	// Block that the request or fulfillment was included in.
+	Block uint64
+}
+
+// Coordinator defines an interface for fetching request and fulfillment metadata from a VRF
+// coordinator.
+type Coordinator interface {
+	// Requests fetches VRF requests that occurred within the specified blocks.
+	Requests(ctx context.Context, fromBlock uint64, toBlock uint64) ([]Event, error)
+
+	// Fulfillments fetches VRF fulfillments that occured since the specified block.
+	Fulfillments(ctx context.Context, fromBlock uint64) ([]Event, error)
+}
+
+// BHS defines an interface for interacting with a BlockhashStore contract.
+type BHS interface {
+	// Store the hash associated with blockNum.
+	Store(ctx context.Context, blockNum uint64) error
+
+	// IsStored checks whether the hash associated with blockNum is already stored.
+	IsStored(ctx context.Context, blockNum uint64) (bool, error)
+}
+
+// NewFeeder creates a new Feeder instance.
+func NewFeeder(
+	logger logger.Logger,
+	coordinator Coordinator,
+	bhs BHS,
+	waitBlocks int,
+	lookbackBlocks int,
+	latestBlock func(ctx context.Context) (uint64, error),
+) *Feeder {
+	return &Feeder{
+		logger:         logger,
+		coordinator:    coordinator,
+		bhs:            bhs,
+		waitBlocks:     waitBlocks,
+		lookbackBlocks: lookbackBlocks,
+		latestBlock:    latestBlock,
+		stored:         make(map[uint64]struct{}),
+		lastRunBlock:   0,
+	}
+}
+
+// Feeder checks recent VRF coordinator events and stores any blockhashes for blocks within
+// waitBlocks and lookbackBlocks that have unfulfilled requests.
+type Feeder struct {
+	logger         logger.Logger
+	coordinator    Coordinator
+	bhs            BHS
+	waitBlocks     int
+	lookbackBlocks int
+	latestBlock    func(ctx context.Context) (uint64, error)
+
+	stored       map[uint64]struct{}
+	lastRunBlock uint64
+}
+
+// Run the feeder.
+func (f *Feeder) Run(ctx context.Context) {
+	latestBlock, err := f.latestBlock(ctx)
+	if err != nil {
+		f.logger.Errorw("Failed to fetch current block number", "error", err)
+		return
+	}
+
+	var (
+		fromBlock        = int(latestBlock) - f.lookbackBlocks
+		toBlock          = int(latestBlock) - f.waitBlocks
+		blockToRequests  = make(map[uint64]map[string]struct{})
+		requestIDToBlock = make(map[string]uint64)
+	)
+	if fromBlock < 0 {
+		fromBlock = 0
+	}
+	if toBlock < 0 {
+		toBlock = 0
+	}
+	reqs, err := f.coordinator.Requests(ctx, uint64(fromBlock), uint64(toBlock))
+	if err != nil {
+		f.logger.Errorw("Failed to fetch VRF requests",
+			"error", err,
+			"latestBlock", latestBlock,
+			"fromBlock", fromBlock,
+			"toBlock", toBlock)
+		return
+	}
+	for _, req := range reqs {
+		if _, ok := blockToRequests[req.Block]; !ok {
+			blockToRequests[req.Block] = make(map[string]struct{})
+		}
+		blockToRequests[req.Block][req.ID] = struct{}{}
+		requestIDToBlock[req.ID] = req.Block
+	}
+
+	fuls, err := f.coordinator.Fulfillments(ctx, uint64(fromBlock))
```

Comment:
I'm unsure of what the latencies of these `Fulfillments` and `Requests` calls look like (perhaps we can measure it in the current BHS), but I wonder if `ctx` at this point doesn't have enough time left in some cases to make the get logs call without timing out. Maybe the BHS feeder timeout in the job should be 60 seconds to account for this.

Reply:
Yeah we might have to play around with the timeout. Are you thinking the default should be 60 seconds? Or we should just configure it that way? I'm guessing 30 will be plenty for ETH mainnet but potentially not for Matic

Reply:
Yeah, I'm mainly concerned about matic, since we could easily get a ton of logs over a block window. I think lets leave that for configuration, but I guess would have to keep in mind that the context-with-timeout would be used for both the fulfillment and the requests log calls.

Reply:
Yup sounds good